mri/scripts/gridsearch.py

```diff
@@ -0,0 +1,188 @@
+                verbose=0, **kwargs):
+    """This function launches off reconstruction for a grid specified
+    through use of kwarg dictionaries.
+
+    Dictionary Convention
+    ---------------------
+    These dictionaries each defined to follow the convention:
+    Each dictionary has a key `init_class` that specifies the
+    initialization class for the operator (exception to
+    this is 'optimizer_kwargs'). Later we have key `args` that holds
+    all the input arguments that can be passed as a keyword dictionary.
+    Each value in this keyword dictionary ,ust be a list of all
+    values you want to search in gridsearch.
+
+    This function finds the search space of parameters and
+    sets up right parameters for '_reconstruct_case' function.
+    Please check the example code for more details.
+
+    Parameters
+    ----------
+    linear_kwargs: dict
+        dictionary for linear operator parameters
+    regularizer_kwargs: dict
+        dictionary for regularizer operator parameters
+    reconstructor_kwargs: dict
+        dictionary for reconstructor operator parameters
+    optimizer_kwargs: dict
+        dictionary for optimizer key word arguments
+    compare_metric_details: dict default None
+        dictionary that holds the metric to be compared and metric
+        direction please refer to `gather_result` documentation.
+    n_jobs: int, default 1
+        number of parallel jobs for execution
+    verbose: int default 0
+        Verbosity level
+        0 => No debug prints
+        1 => View best results if present
+
+    **kwargs: keyword arguments that are passes to reconstruction
+        these holds extra arguments that are passed to
+        '_reconstruct_case' function.
+        important arguments are 'ksapce_data' and 'fourier_kwargs'
+    """
+    # Convert non-list elements to list so that we can create
+    # search space
+    all_reformatted_kwargs = []
+    for specific_kwargs in [linear_kwargs, regularizer_kwargs,
+                            reconstructor_kwargs, optimizer_kwargs]:
+        for key, value in specific_kwargs['args'].items():
+            if not isinstance(value, list) and \
+                    not isinstance(value, np.ndarray):
+                specific_kwargs['args'][key] = [value]
+        all_reformatted_kwargs.append(specific_kwargs)
+    # Create Search space
+    cross_product_list = list(itertools.product(*tuple(sum(
+        [list(all_reformatted_kwargs[i]['args'].values())
+         for i in range(len(all_reformatted_kwargs))],
+        []
+    ))))
+    # Obtain Initialization classes
+    init_classes = [
+        all_reformatted_kwargs[i]['init_class']
+        for i in range(len(all_reformatted_kwargs)-1)
+    ]
+    # Obtain key names for all cases
+    key_names = list([
+        list(all_reformatted_kwargs[i]['args'].keys())
+        for i in range(len(all_reformatted_kwargs))
+    ])
+    reshaped_cross_product = []
```

Comment:
I am not sure I understand what's happening here.

Reply:
It is just reshaping the arguments same as key_names, this makes the _reconstruct_case function more readable

Reply:
can you detail a bit this part, like have it in several lines with variable names that explain what's happening (with comments potentially).

Reply:
Done, However, I think I just added more code, let me know if it can be made better

Reply:
why do you need to do `*tuple(linear_params['kwargs'].values())` and not just `*linear_params['kwargs'].values()`?

Reply:
Oh yeah, tuple is no longer needed.